wardrobe/tradable_kitties/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+        KittyHelpers::can_breed(&mom.kitty_basic_data, &dad.kitty_basic_data)?;
+        Ok(())
+    }
+
+    fn check_new_family(
+        mom: &TradableKittyData,
+        dad: &TradableKittyData,
+        new_tradable_kitty_family: &[DynamicallyTypedData],
+    ) -> Result<(), Self::Error> {
+        let new_tradable_kitty_mom = TradableKittyData::try_from(&new_tradable_kitty_family[0])?;
+        let new_tradable_kitty_dad = TradableKittyData::try_from(&new_tradable_kitty_family[1])?;
+        let new_tradable_kitty_child = TradableKittyData::try_from(&new_tradable_kitty_family[2])?;
+
+        let new_basic_kitty_mom: DynamicallyTypedData =
+            new_tradable_kitty_mom.kitty_basic_data.into();
+        let new_basic_kitty_dad: DynamicallyTypedData =
+            new_tradable_kitty_dad.kitty_basic_data.into();
+        let new_basic_kitty_child: DynamicallyTypedData =
+            new_tradable_kitty_child.kitty_basic_data.into();
+
+        let mut new_family: Vec<DynamicallyTypedData> = Vec::new();
+        new_family.push(new_basic_kitty_mom);
+        new_family.push(new_basic_kitty_dad);
+        new_family.push(new_basic_kitty_child);
+
+        KittyHelpers::check_new_family(&mom.kitty_basic_data, &dad.kitty_basic_data, &new_family)?;
+        Ok(())
+    }
+}
+
+impl<const ID: u8> Buy for TradableKittyHelpers<ID> {
+    type Error = TradableKittyConstraintCheckerError;
+    fn can_buy(
+        input_data: &[DynamicallyTypedData],
+        output_data: &[DynamicallyTypedData],
+    ) -> Result<(), Self::Error> {
+        let mut input_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+        let mut output_coin_data: Vec<DynamicallyTypedData> = Vec::new();
+        let mut input_kitty_data: Vec<DynamicallyTypedData> = Vec::new();
+        let mut output_kitty_data: Vec<DynamicallyTypedData> = Vec::new();
+
+        let mut total_input_amount: u128 = 0;
+        let mut total_price_of_kitty: u128 = 0;
+
+        for utxo in input_data {
+            if let Ok(coin) = utxo.extract::<Coin<ID>>() {
+                let utxo_value = coin.0;
+                ensure!(
+                    utxo_value > 0,
+                    TradableKittyConstraintCheckerError::ZeroValueCoin
+                );
+                input_coin_data.push(utxo.clone());
+                total_input_amount = total_input_amount
+                    .checked_add(utxo_value)
+                    .ok_or(TradableKittyConstraintCheckerError::ValueOverflow)?;
+
+                // Process Kitty
+            } else if let Ok(tradable_kitty) = utxo.extract::<TradableKittyData>() {
+                if !tradable_kitty.is_available_for_sale {
+                    return Err(Self::Error::KittyNotForSale);
+                }
+                let price = match tradable_kitty.price {
+                    None => return Err(Self::Error::KittyPriceCantBeNone),
+                    Some(p) => p,
+                };
```

Comment:
If `price` is already an option, then why do you also need a `is_available_for_sale` field. Can't it just be that if the price is `None` then is it not for sale.

Reply:
I agree, will remove is_available_for_sale field.